chalk-solve/src/infer/unify.rs

```diff
@@ -333,6 +333,87 @@ impl<'t, I: Interner> Unifier<'t, I> {
         }
     }
 
+    fn unify_const_const<'a>(&mut self, a: &'a Const<I>, b: &'a Const<I>) -> Fallible<()> {
+        let interner = self.interner;
+
+        if let Some(n_a) = self.table.normalize_const_shallow(interner, a) {
+            return self.unify_const_const(&n_a, b);
+        } else if let Some(n_b) = self.table.normalize_const_shallow(interner, b) {
+            return self.unify_const_const(a, &n_b);
+        }
+
```

Comment:
I was puzzled by this construction as well, but forgot about it. I guess it would make sense if you can bind an inference var to an inference var (as opposed to unifying them), but that shouldn't happen as far as I know. @nikomatsakis what do you think?

Reply:
You can do the `unwrap_or`. I'm not sure why I wrote it that way tbh. I guess I was just thinking "functionally" -- i.e,. of an inference rule like

```
Bound(V) = T'
Unify(T', T)
-----------------
Unify(V, T)
```